internal/component/database_observability/mysql/collector/schema_table.go

```diff
@@ -432,30 +434,48 @@ func (c *SchemaTable) fetchColumnsDefinitions(ctx context.Context, schemaName st
 	defer idxRS.Close()
 
 	for idxRS.Next() {
-		var indexName, columnName, indexType string
+		var indexName, indexType string
 		var seqInIndex, nonUnique int
-		var nullable sql.NullString
-		if err := idxRS.Scan(&indexName, &seqInIndex, &columnName, &nullable, &nonUnique, &indexType); err != nil {
+		var columnName, expression, nullable sql.NullString
+		if err := idxRS.Scan(&indexName, &seqInIndex, &columnName, &expression, &nullable, &nonUnique, &indexType); err != nil {
 			level.Error(c.logger).Log("msg", "failed to scan table indexes", "schema", schemaName, "table", tableName, "err", err)
 			return nil, err
 		}
 
+		// mysql docs describe column and expression as mutually exclusive,
+		// but at least one of them must be present.
+		if !columnName.Valid && !expression.Valid {
+			level.Error(c.logger).Log("msg", "index without a column or expression", "schema", schemaName, "table", tableName, "index", indexName)
+			continue
+		}
+
 		// Append column to the last index if it's the same as the previous one (i.e. multi-column index)
 		if nIndexes := len(tblSpec.Indexes); nIndexes > 0 && tblSpec.Indexes[nIndexes-1].Name == indexName {
 			lastIndex := &tblSpec.Indexes[nIndexes-1]
-			if len(lastIndex.Columns) != seqInIndex-1 {
-				level.Error(c.logger).Log("msg", "unexpected index column sequence", "schema", schemaName, "table", tableName, "index", indexName, "column", columnName)
+			if len(lastIndex.Columns)+len(lastIndex.Expressions) != seqInIndex-1 {
+				level.Error(c.logger).Log("msg", "unexpected index ordinal position", "schema", schemaName, "table", tableName, "index", indexName, "seq", seqInIndex, "len_columns", len(lastIndex.Columns), "len_expressions", len(lastIndex.Expressions))
 				continue
 			}
-			lastIndex.Columns = append(lastIndex.Columns, columnName)
+
+			if columnName.Valid {
+				lastIndex.Columns = append(lastIndex.Columns, columnName.String)
+			} else if expression.Valid {
```

Comment:
What if you have both a columnName and expression?

Reply:
Docs says they are mutually exclusive, and there's no syntax in current mysql versions to create an index with both a column and an expression. We could add support for such a possible edge case, but I'm thinking it might be better to do it when/if we find anything odd (just to avoid supporting all possible edge cases as it makes the code harder to follow).